en/cpp/api_changes.md

```diff
@@ -21,9 +21,91 @@ This means that breaking changes to the API result in a bump of the major versio
   We plan to release version 1.0.0 in the near future, after which the above strategy will be adopted (i.e. currently breaking changes can occur in both minor and patch releases).
 
 
+## v0.33.0
+
+There are some changes in how systems are discovered, connected and accessed with this release.
+The previous APIs are still available but marked deprecated and they will be removed in the future.
+The deprecations are also implemented as compiler warnings.
+
+### Discovery of systems
+
+The 64bit (8 bytes) uuid is deprecated because it needed to be extended to 18 bytes.
+For more info, see this [MAVLink pull request](https://github.com/mavlink/mavlink/pull/786).
+Therefore, the discovery no longer contains the uuid.
+If the uid is required it can be found in the [Info plugin](../api_reference/classmavsdk_1_1_info.md#classmavsdk_1_1_info_1a812ed66265b7427bc781faec3f0fa89e).
+
+**Previous way of discovering systems:**
+```cpp
+std::promise<void> discover_promise;
+auto discover_future = discover_promise.get_future();
+
+mavsdk.register_on_discover([&discover_promise](uint64_t uuid) {
+    std::cout << "Discovered system with UUID: " << uuid << std::endl;
+    discover_promise.set_value();
+});
+
+discover_future.wait();
+```
+
+**New way of discovering a system:**
+```cpp
+std::promise<void> discover_promise;
+auto discover_future = discover_promise.get_future();
+
+mavsdk.subscribe_on_change([&mavsdk, &discover_promise]() {
+    const auto system = mavsdk.systems().at(0);
```

Comment:
The `.at(0)` is a bit confusing for two reasons:
1. I think this is because systems is now a shared point to a system, and you're grabbing the first system. But at this point you don't know about that - perhaps a forward reference to next section.
2. What if you're grabbing the second system? Does the user have any way to grab the system that actually got added?

Reply:
1. Oh, you're right. Maybe a comment like `// see how to access systems below`?

Reply:
1. I don't get that, sorry :see_no_evil:. `mavsdk.systems()` gives you the list of systems, and because you know it just changed, you can be fairly sure that there is one system and that it is connected. I don't know if it is a feature or not, but I believe that we never remove any `System` from the `systems()` list, so the first one discovered will always be at [0] (it's not like the list will get shuffled or something).
2. I believe that whether or not you get an indication of which system changed will not matter in your implementation in this function. Right now, anyway `subscribe_on_change` is used both for discovery and for checking the connection state, so you will have this logic anyway. I could imagine separating them by having another callback as part of a `System`, though :thinking::

```c++
mavsdk.subscribe_on_discover(<callback>); // This tells me that there is a new system in the `systems()` list
system.subscribe_on_change(<callback>); // This tells me that this particular system's state has changed
```

Would that make sense? Otherwise `systems()` as we have it now is also fine with me, but I don't think that the id brings value :+1:.

Reply:
1. This is only a doc thing. We are using the new API but it gets explained further below in the docs.
2. `mavsdk.subscribe_on_discover` so this would trigger every time that a new system is added, and they will never be removed because, ever.
And then I would maybe call it `system_subscribe_on_connection_change` but then you still don't know if it has just connected or disconnected.

Reply:
> but then you still don't know if it has just connected or disconnected

Right. If the only state that can change is `is_connected()`, then it could also be something like `system.subscribe_is_connected([](bool is_connected) { ... });` And if we think that there could be other things in the state, then `subscribe_on_connection_change` tells you that you may want to check `system.is_connected()`, which seems fine to me :+1:.

A bonus from splitting them is that then we don't care about whether or not the `systems()` list changes (so later we could even remove systems if we wanted to): `subscribe_on_discover` tells you that you will find a new system in the list, and if you are interested you can get a pointer to it, but after that you will never need to access this system through that list again (because you just keep the pointer).

Reply:
`system.subscribe_is_connected([](bool is_connected) { ... });` makes sense, perfect.

So I don't understand your pointer idea exactly. Can you put that in code?

Reply:
```c++
// Discovery step
auto discover_prom = std::promise<std::shared_ptr<System>>{};
auto discover_fut = discover_prom.get_future();
mavsdk.subscribe_on_discover([&mavsdk, &prom]() {
    prom.set_value(mavsdk.systems().at(0));
    mavsdk.subscribe_on_discover(nullptr);
});

auto system = discover_fut.get();

// From this point on, I will never need to call `mavsdk.systems()` again,
// as I will just work with my `system`. So I don't care if the `systems()`
// list gets reshuffled from time to time.
system->subscribe_is_connected([](bool is_connected) {
    if (!is_connected) {
        std::cout << "ERROR: system got disconnected, which is not supported by this script" << std::endl;
        exit(1);
    }
});

// Create my plugins and use them
Action action(system);
...
```